### PR TITLE
Adds list command of storage for devfile v2

### LINKED ISSUE #3704: odo storage list should list all volumes from devfile components
/kind user-story

## User Story
As a developer, I want to list the storage defined for the component so that I can have an idea about the storage which are defined for the component.

## Acceptance Criteria 
- [x] It should list the storage with proper states on `odo storage list`.
- [x] It should have json output.

## Links 
- Feature Request: 
- Related issue: 

/kind user-story

### PR DESCRIPTION
**Pull request by mik-dass**
**What type of PR is this?**

/kind feature

**What does does this PR do / why we need it**:

It adds `odo storage list` command for devfile volumes

**Which issue(s) this PR fixes**:

Fixes #3704 

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer**:

TBA

### CODE REVIEW

File: pkg/odo/cli/storage/list.go [hunk 1]
```diff
@@ -2,7 +2,13 @@ package storage
 
 import (
 	"fmt"
+	"github.com/openshift/odo/pkg/devfile"
+	devfileParser "github.com/openshift/odo/pkg/devfile/parser"
```
Comment (dev-gaur) [inline pkg/odo/cli/storage/list.go:6]: camel cases shouldn't be used in golang.

File: pkg/odo/cli/storage/list.go [hunk 2]
```diff
@@ -2,7 +2,13 @@ package storage
...
+	"github.com/openshift/odo/pkg/devfile/parser/data/common"
+	"github.com/openshift/odo/pkg/odo/cli/component"
+	odoUtil "github.com/openshift/odo/pkg/util"
```
Comment (dev-gaur) [inline pkg/odo/cli/storage/list.go:9]: same.

File: pkg/odo/cli/storage/list.go (hunk 1 above)
Comment (dev-gaur) [reply pkg/odo/cli/storage/list.go:6]: I mean, for package names and aliases.

**Review by dev-gaur — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/storage/list.go (hunk 1 above)
Comment (mik-dass) [reply pkg/odo/cli/storage/list.go]: fixed

File: pkg/odo/cli/storage/list.go (hunk 2 above)
Comment (mik-dass) [reply pkg/odo/cli/storage/list.go]: fixed

File: pkg/storage/storage.go [hunk 3]
```diff
@@ -550,3 +567,127 @@ func MachineReadableSuccessOutput(storageName string, message string) {
...
+		if !found {
+			return StorageList{}, fmt.Errorf("mount path for pvc %s not found", pvc.Name)
+		}
+	}
+
+	return StorageList{Items: storage}, nil
+}
+
+// GetLocalDevfileStorage lists the storage from the devfile
+func GetLocalDevfileStorage(devfileData data.DevfileData) StorageList {
```
Comment (adisky) [inline pkg/storage/storage.go:623]: it can be a private function

File: pkg/storage/storage.go [hunk 4]
```diff
@@ -550,3 +567,127 @@ func MachineReadableSuccessOutput(storageName string, message string) {
 
 	machineoutput.OutputSuccess(machineOutput)
 }
+
+// DevfileListMounted lists the storage which are mounted on a container
+func DevfileListMounted(kClient *kclient.Client, componentName string) (StorageList, error) {
```
Comment (adisky) [inline pkg/storage/storage.go:572]: it could be a private function

**Review by adisky — COMMENTED** [review]
for no storage, it is printing empty string in place of component name.
$ odo storage list
The component '' has no storage attached

**Comment by mik-dass** [comment]
@adisky Fixed.

**Comment by mik-dass** [comment]
```
odo]  ✗  Command 'tar xf - -C /tmp/src' in container failed.
[odo] 
[odo]  ✗  stdout: 
[odo] 
[odo]  ✗  stderr: tar: /tmp/src: Cannot open: No such file or directory
[odo] tar: Error is not recoverable: exiting now
[odo] 
[odo] 
[odo]  ✗  err: error while streaming command: command terminated with exit code 2
[odo] 
[odo]  ✗  command terminated with exit code 2
[odo]  ✗  Syncing files to the component [991ms]
```
/retest

**Comment by kadel** [comment]
I can see two storage listed even if there is only one volume in devfile
```
▶ odo storage list
The component 'java-maven' has the following storage attached:
NAME     SIZE     PATH               CONTAINER     STATE
m2       1Gi      /home/user/.m2     tools         Not Pushed
m2       5Gi      /home/user/.m2     tools         Locally Deleted
```

```yaml
schemaVersion: 2.0.0
metadata:
  name: java-maven
  version: 1.0.0
components:
  - container:
      name: tools
      image: quay.io/eclipse/che-java11-maven:nightly
      memoryLimit: 512Mi
      mountSources: true
      endpoints:
        - name: 'http-8080'
          targetPort: 8080
          exposure: public
      volumeMounts:
        - name: m2
          path: /home/user/.m2
  - volume:
      name: m2
commands:
  - exec:
      id: mvn-package
      component: tools
      commandLine: "mvn package"
      group:
        kind: build
        isDefault: true
  - exec:
      id: run
      component: tools
      commandLine: "java -jar target/*.jar"
      group:
        kind: run
        isDefault: true
  - exec:
      id: debug
      component: tools
      commandLine: "java -Xdebug -Xrunjdwp:server=y,transport=dt_socket,address=${DEBUG_PORT},suspend=n -jar target/*.jar"
      group:
        kind: debug
        isDefault: true
```

**Comment by mik-dass** [comment]
> I can see two storage listed even if there is only one volume in devfile

Do you have a volume of 5Gi size on the cluster?

**Comment by dharmit** [comment]
State should be `Pushed`, no?
```sh
$ odo storage create mystorage --path /opt/mystorage --size=1Gi
 ✓  Added storage mystorage to node-todo

Please use `odo push` command to make the storage accessible to the component

$ odo push
Validation
 ✓  Checking component [16ms]

Configuration changes
 ✓  Retrieving component data [14ms]
 ✓  Applying configuration [18ms]

Applying URL changes
 ✓  URLs are synced with the cluster, no changes are required.

Pushing to component node-todo of type local
 ✓  Checking file changes for pushing [443603ns]
 ✓  Waiting for component to start [5ms]
 ✓  Syncing files to the component [113ms]
 ✓  Building component [2s]
 ✓  Changes successfully pushed to component

$ odo storage list
The component 'node-todo' has the following storage attached:
NAME          SIZE     PATH               STATE
mystorage     1Gi      /opt/mystorage     Not Pushed
```

**Comment by kadel** [comment]
> A similar nodejs devfile works for me locally. Do you have a volume of 5Gi size on the cluster?

I tried again in a clean project, and the result is the same.
Interestingly `odo storage list` before pushing is ok, but after the push I see two volumes. 
Could this have anything to do with the default size?
Storage backend can have minimal allowed size. So even if you request PVC of size 1Gi, it might create 5Gi PVC.





```
▶ odo project create test-new
 ✓  Project 'test-new' is ready for use
 ✓  New project created and now using project: test-new

▶ odo create java-maven
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile existence [57768ns]
 ✓  Checking devfile compatibility [106579ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [98459ns]
 ✓  Validating devfile component [145213ns]

Please use `odo push` command to create the component with source deployed

▶ odo storage list
The component 'java-maven' has the following storage attached:
NAME     SIZE     PATH               STATE
m2       1Gi      /home/user/.m2     Not Pushed


▶ odo push

Validation
 ✓  Validating the devfile [111524ns]

Creating Kubernetes resources for component java-maven
 ✓  Waiting for component to start [19s]

Applying URL changes
 ✓  URL http-8080: http://http-8080-java-maven-test-new.apps.testocp4x.psiodo.net/ created

Syncing to component java-maven
 ✓  Checking files for pushing [6ms]
 ✓  Syncing files to the component [1m]

Executing devfile commands for component java-maven
 ✓  Executing mvn-package command "mvn package", if not running [19s]
 ✓  Executing run command "java -jar target/*.jar", if not running [2s]

Pushing devfile component java-maven
 ✓  Changes successfully pushed to component

▶ odo storage list
The component 'java-maven' has the following storage attached:
NAME     SIZE     PATH               CONTAINER     STATE
m2       1Gi      /home/user/.m2     tools         Not Pushed
m2       5Gi      /home/user/.m2     tools         Locally Deleted
```

**Comment by dharmit** [comment]
My bad about https://github.com/openshift/odo/pull/3788#issuecomment-682434606. I didn't realize I was using s2i based component. `odo storage list` works just fine when I switch to experimental mode and use devfile component instead:
```
$ odo storage list
The component 'node-todo' has the following storage attached:
NAME           SIZE     PATH                STATE
mystorage      1Gi      /opt/mystorage      Pushed
mystorage2     1Gi      /opt/mystorage2     Pushed
```

**Comment by mik-dass** [comment]
@kadel Fixed

**Comment by kadel** [comment]
> @kadel Fixed

/approve

**Pull request merged**